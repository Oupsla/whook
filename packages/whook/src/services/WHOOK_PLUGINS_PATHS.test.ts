import initWhookPluginsPaths from './WHOOK_PLUGINS_PATHS';
import YError from 'yerror';

describe('WHOOK_PLUGINS_PATHS', () => {
  const log = jest.fn();
  const resolve = jest.fn();

  beforeEach(() => {
    log.mockReset();
    resolve.mockReset();
  });

  describe('should work', () => {
    it('with no plugin at all', async () => {
      const WHOOK_PLUGINS_PATHS = await initWhookPluginsPaths({
        PROJECT_SRC: '/home/whoami/my-whook-project/src',
        WHOOK_PLUGINS: [],
        resolve: (resolve as unknown) as RequireResolve,
        log,
      });

      expect({
        WHOOK_PLUGINS_PATHS,
        logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
        resolveCalls: resolve.mock.calls,
      }).toMatchSnapshot();
    });

    it('with some plugins', async () => {
      resolve.mockImplementationOnce(
        () => '/var/lib/node/node_modules/@whook/cli',
      );
      resolve.mockImplementationOnce(
        () => '/var/lib/node/node_modules/@whook/whook',
      );

      const WHOOK_PLUGINS_PATHS = await initWhookPluginsPaths({
        PROJECT_SRC: '/home/whoami/my-whook-project/src',
        WHOOK_PLUGINS: ['@whook/cli', '@whook/whook'],
        resolve: (resolve as unknown) as RequireResolve,
        log,
      });

      expect({
        WHOOK_PLUGINS_PATHS,
        logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
        resolveCalls: resolve.mock.calls,
      }).toMatchSnapshot();
    });
  });

  describe('should fail', () => {
    it('with unexisting plugin', async () => {
      resolve.mockImplementationOnce(() => {
        throw new YError('E_NO_MODULE');
      });

      try {
        await initWhookPluginsPaths({
          PROJECT_SRC: '/home/whoami/my-whook-project/src',
          WHOOK_PLUGINS: ['@whook/unreal'],
          log,
          resolve: (resolve as unknown) as RequireResolve,
        });

        throw new YError('E_UNEXPECTED_SUCCESS');
      } catch (err) {
        expect({
          errorCode: err.code,
          errorParams: err.params,
          logCalls: log.mock.calls.filter(args => 'stack' !== args[0]),
          resolveCalls: resolve.mock.calls,
        }).toMatchSnapshot();
      }
    });
  });
});
