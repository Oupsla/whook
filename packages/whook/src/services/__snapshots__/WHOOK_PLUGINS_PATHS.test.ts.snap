// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WHOOK_PLUGINS_PATHS should fail with unexisting plugin 1`] = `
Object {
  "errorCode": "E_BAD_WHOOK_PLUGIN",
  "errorParams": Array [
    "@whook/unreal",
  ],
  "logCalls": Array [],
  "resolveCalls": Array [
    Array [
      "@whook/unreal",
      Object {
        "paths": Array [
          "/home/whoami/my-whook-project/src",
        ],
      },
    ],
  ],
}
`;

exports[`WHOOK_PLUGINS_PATHS should work with no plugin at all 1`] = `
Object {
  "WHOOK_PLUGINS_PATHS": Array [],
  "logCalls": Array [],
  "resolveCalls": Array [],
}
`;

exports[`WHOOK_PLUGINS_PATHS should work with some plugins 1`] = `
Object {
  "WHOOK_PLUGINS_PATHS": Array [
    "/var/lib/node/node_modules/@whook",
    "/var/lib/node/node_modules/@whook",
  ],
  "logCalls": Array [
    Array [
      "debug",
      "➰ - Plugin \\"@whook/cli\\" resolved to: /var/lib/node/node_modules/@whook",
    ],
    Array [
      "debug",
      "➰ - Plugin \\"@whook/whook\\" resolved to: /var/lib/node/node_modules/@whook",
    ],
  ],
  "resolveCalls": Array [
    Array [
      "@whook/cli",
      Object {
        "paths": Array [
          "/home/whoami/my-whook-project/src",
        ],
      },
    ],
    Array [
      "@whook/whook",
      Object {
        "paths": Array [
          "/home/whoami/my-whook-project/src",
        ],
      },
    ],
  ],
}
`;
